DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/src/fp_pragmas.cpp

```diff
@@ -0,0 +1,269 @@
+void enable_reassociate(const device_selector &selector, const double a,
+                        const double b, const double c, const double d,
+                        double &res) {
+  try {
+    queue q(selector, dpc_common::exception_handler,
+            property::queue::enable_profiling{});
+
+    buffer<double, 1> bufferA(&a, 1);
+    buffer<double, 1> bufferB(&b, 1);
+    buffer<double, 1> bufferC(&c, 1);
+    buffer<double, 1> bufferD(&d, 1);
+    buffer<double, 1> bufferE(&res, 1);
+
+    event e = q.submit([&](handler &h) {
+      accessor accessorA(bufferA, h, read_only);
+      accessor accessorB(bufferB, h, read_only);
+      accessor accessorC(bufferC, h, read_only);
+      accessor accessorD(bufferD, h, read_only);
+      accessor accessorRes(bufferE, h, write_only, noinit);
+
+      // The "kernel_args_restrict" tells the compiler that a, b, and r
+      // do not alias. For a full explanation, see:
+      //    DPC++FPGA/Tutorials/Features/kernel_args_restrict
+      h.single_task<ReassociateOnKernel>([=]() [[intel::kernel_args_restrict]] {
+
+      // Enables the compiler to reorder floating-point operations to improve
+      // performance and on-chip area.
+#pragma clang fp reassociate(on)
+        accessorRes[0] =
+            accessorA[0] + accessorB[0] + accessorC[0] + accessorD[0];
+      });
+    });
+  } catch (sycl::exception const &e) {
+    // Catches exceptions in the host code
+    std::cerr << "Caught a SYCL host exception:\n" << e.what() << "\n";
+
+    // Most likely the runtime couldn't find FPGA hardware!
+    if (e.get_cl_code() == CL_DEVICE_NOT_FOUND) {
+      std::cerr << "If you are targeting an FPGA, please ensure that your "
+                   "system has a correctly configured FPGA board.\n";
+      std::cerr << "Run sys_check in the oneAPI root directory to verify.\n";
+      std::cerr << "If you are targeting the FPGA emulator, compile with "
+                   "-DFPGA_EMULATOR.\n";
+    }
+    std::terminate();
+  }
+}
+
+int main() {
+  const double a = 1.0, b = 2.0, c = 3.0, d = 4.0;
+  // Result variables
+  double res_contract_off, res_reassociate_off, res_contract_fast,
+      res_reassociate_on;
+  const double golden_res_contract = 21.0;
+  const double golden_res_reassociate = 10.0;
+
+  // Run the kernel on either the FPGA emulator, or FPGA
+#if defined(FPGA_EMULATOR)
+  INTEL::fpga_emulator_selector selector;
+#else
+  INTEL::fpga_selector selector;
+#endif
+
+  // Test fp contract pragma
+  disable_contract(selector, a, b, c, d, res_contract_off);
+  enable_contract(selector, a, b, c, d, res_contract_fast);
+  // Test fp reassociate pragma
+  disable_reassociate(selector, a, b, c, d, res_reassociate_off);
+  enable_reassociate(selector, a, b, c, d, res_reassociate_on);
+
+  // Using fp contract and fp reassociate pragmas will cause some nuances in
+  // precision due to rounding or reordering of math operations, but the results
+  // should be close to the strict results.
```

Comment:
How about this:

The fp contract and reassociate pragmas can save area and improve performance, but because the operations can be reordered we may get slightly different results from those computed by the CPU.  The EPISILON constant allows us to define what we consider to be 'close enough' to an exact match.